dbfit-java/derby/src/main/java/dbfit/environment/DerbyEnvironment.java

```diff
@@ -82,9 +84,95 @@ public Pattern getParameterPattern() {
         }
     }
 
+    private boolean isProcedureOrFunction(String schemaName, String objectName,
+                                                             boolean checkForFunction)
+                                                                   throws SQLException {
+        DatabaseMetaData dbmd = getConnection().getMetaData();
+        try (ResultSet rs = checkForFunction ? dbmd.getFunctions(null, schemaName, objectName) :
+                                               dbmd.getProcedures(null, schemaName, objectName)) {
+            return rs.next();
+        }
+    }
+
+    private Direction getParamDirection(int columnType, boolean isFunction) {
+        Direction dir = null;
+        if (isFunction) {
+            switch (columnType) {
+                case DatabaseMetaData.functionColumnIn:
+                    dir = Direction.INPUT;
+                    break;
+                case DatabaseMetaData.functionReturn:
+                    dir = Direction.RETURN_VALUE;
+                    break;
+            }
+        } else {
+            switch (columnType) {
+                case DatabaseMetaData.procedureColumnIn:
+                    dir = Direction.INPUT;
+                    break;
+                case DatabaseMetaData.procedureColumnInOut:
+                    dir = Direction.INPUT_OUTPUT;
+                    break;
+                case DatabaseMetaData.procedureColumnOut:
+                    dir = Direction.OUTPUT;
+                    break;
+            }
+        }
+        if (dir == null) {
+            throw new UnsupportedOperationException(
+                "Type " + columnType + " is not a supported type for " +
+                (isFunction ? "function" : "procedure") + " parameters");
+        }
+        return dir;
+    }
+
+    private Map<String, DbParameterAccessor> getParamMap(String schemaName,
+                                                         String procedureName,
+                                                         boolean isFunction)
+                                                         throws SQLException {
+        DatabaseMetaData dbmd = getConnection().getMetaData();
+        int position = 0;
+        Direction paramDirection;
+        String dataType, paramName;
+        Map<String, DbParameterAccessor> allParams = new HashMap<>();
+        ResultSet rs = isFunction ? dbmd.getFunctionColumns(null, schemaName, procedureName, "%") :
+                                    dbmd.getProcedureColumns(null, schemaName, procedureName, "%");
+        while (rs.next()) {
+            paramName = defaultIfNull(rs.getString("COLUMN_NAME"), "");
```

Comment:
The details of unwrapping a of single parameter into accessor is a candidate to be extracted.